Suggest class name in factory instantiation

## Intro
Instead of using directly the classes from C++, we use factories (see [here](https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/util/factory.h)). So for example in Python, instead of
```python
import shogun as sg
kernel = sg.GaussianKernel()
```
we have
```python
import shogun as sg
kernel = sg.kernel("GaussianKernel")
```
To find out what all possible names you can have a look at `src/shogun/base/class_list.cpp`, after compiling shogun.
The issue with this approach is that it is easy to mistype a name, e.g. `"Gaussiankernel"`, and after several hours of writing code these errors might happen more frequently and not so obvious to debug.
Right now you would get a message like this:
```python
SystemError: [ERROR] In file /home/gil/shogun/src/shogun/base/class_list.h line 51: Class Gaussiankernel with primitive type SGOBJECT does not exist.
```
## The task
Catch these errors with SWIG, for example in src/interfaces/shogun.i and try to find out what the user actually meant. This should be written in C++, which would make it accessible to all languages.

## Some pointers
- A good approach would be to calculate the distance between the string the user passed and all the classes in shogun
- All classes in shogun that are exposed via factories can be accessed with [shogun::available_objects()](https://github.com/shogun-toolbox/shogun/blob/e3e8cd197b4dde5a6c75136bea54e7fdc141a947/src/shogun/base/class_list.h#L71)
- A good distance metric for this is levenshtein, which is not implemented anywhere in shogun, but it is pretty straighforward (see [here](https://rosettacode.org/wiki/Levenshtein_distance#C.2B.2B)).
- Check #4473, it has a Python implementation of this

--- a/src/shogun/base/base_types.h
+++ b/src/shogun/base/base_types.h
@@ -167,6 +167,20 @@ namespace shogun
 
 	template <typename T>
 	inline constexpr bool is_auto_value_v = is_auto_value<T>::value;
+
+
+	template<typename classType>
+	std::string_view name_lookup()
+	{
+		if constexpr(std::is_same_v<classType, Machine>)
+		{
+			return "Machine";
+		}
+		else 
+		{
+			return "SGObject";
+		}
+	};
 } // namespace shogun
 
 #endif // BASE_TYPES__H

--- a/src/shogun/base/class_list.cpp.templ
+++ b/src/shogun/base/class_list.cpp.templ
@@ -75,11 +75,11 @@ std::string shogun::find_correct_name(const std::string& name)
 {
     using pair = std::pair<int32_t, std::string>;
     std::priority_queue<pair, std::vector<pair>, std::greater<pair>> names;
-    LevenshteinDistance levenshtein;
+    auto levenshtein = std::make_shared<LevenshteinDistance>();
     std::for_each(classes.begin(), classes.end(), 
     	[&](const auto& class_pair){
 			auto class_name = class_pair.first;
-			auto dis = levenshtein.distance(class_name, name);
+			auto dis = levenshtein->distance(class_name, name);
 			names.push(std::make_pair(dis, class_name));
     	});
     return names.top().second;

--- a/src/shogun/base/class_list.cpp.templ
+++ b/src/shogun/base/class_list.cpp.templ
@@ -70,3 +70,17 @@ std::set<std::string> shogun::available_objects()
 	}
 	return result;
 }
+
+std::string shogun::find_correct_name(const std::string& name)
+{
+    using pair = std::pair<int32_t, std::string>;
+    std::priority_queue<pair, std::vector<pair>, std::greater<pair>> names;
+    LevenshteinDistance levenshtein;
+    std::for_each(classes.begin(), classes.end(), 
+    	[&](const auto& class_pair){
+			auto class_name = class_pair.first;
+			auto dis = levenshtein.distance(class_name, name);
+			names.push(std::make_pair(dis, class_name));
+    	});
+    return names.top().second;
+}

--- a/src/shogun/base/class_list.h
+++ b/src/shogun/base/class_list.h
@@ -17,6 +17,7 @@
 
 #include <set>
 #include <string>
+#include <queue>
 
 namespace shogun {
 	class SGObject;
@@ -33,6 +34,9 @@ namespace shogun {
 	 * the requested type and the object's type do not match.
 	 *
 	 */
+
+	std::string find_correct_name(const std::string& name);
+
 	template <class T>
 	std::shared_ptr<T> create_object(
 	    const char* name,
@@ -41,9 +45,9 @@ namespace shogun {
 		auto object = create(name, pt);
 		if (!object)
 		{
+			const auto& correct_name = find_correct_name(name);
 			error(
-			    "Class {} with primitive type {} does not exist.", name,
-			    ptype_name(pt).c_str());
+			    "Class {} does not exist. Did you mean {} ?", name, correct_name);
 		}
 		auto cast = std::dynamic_pointer_cast<T>(object);
 		if (cast == nullptr)

--- a/src/shogun/base/class_list.h
+++ b/src/shogun/base/class_list.h
@@ -28,15 +28,17 @@ namespace shogun {
 	 */
 	std::shared_ptr<SGObject> create(const char* sgserializable_name, EPrimitiveType generic);
 
+	/** Use Levenshtein distance to find the most similar name
+	 *
+	 */
+	std::string find_correct_name(const std::string& name);
+
 	/** Creates new shogun instance, typed.
 	 *
 	 * Throws an exception in case there is no such classname or
 	 * the requested type and the object's type do not match.
 	 *
 	 */
-
-	std::string find_correct_name(const std::string& name);
-
 	template <class T>
 	std::shared_ptr<T> create_object(
 	    const char* name,

--- a/src/shogun/base/class_list.h
+++ b/src/shogun/base/class_list.h
@@ -22,6 +22,18 @@
 namespace shogun {
 	class SGObject;
 
+	template<typename classType>
+	std::string_view name_lookup()
+	{
+		if constexpr(std::is_same_v<classType, Machine>)
+		{
+			return "Machine";
+		}
+		else 
+		{
+			return "SGObject";
+		}
+	};
 	/** new shogun instance
 	 * @param sgserializable_name
 	 * @param generic
@@ -49,7 +61,7 @@ namespace shogun {
 		{
 			const auto& correct_name = find_correct_name(name);
 			error(
-			    "Class {} does not exist. Did you mean {} ?", name, correct_name);
+			    "{} {} does not exist. Did you mean {} ?", name_lookup<T>(), name, correct_name);
 		}
 		auto cast = std::dynamic_pointer_cast<T>(object);
 		if (cast == nullptr)

--- a/src/shogun/base/class_list.h
+++ b/src/shogun/base/class_list.h
@@ -22,18 +22,6 @@
 namespace shogun {
 	class SGObject;
 
-	template<typename classType>
-	std::string_view name_lookup()
-	{
-		if constexpr(std::is_same_v<classType, Machine>)
-		{
-			return "Machine";
-		}
-		else 
-		{
-			return "SGObject";
-		}
-	};
 	/** new shogun instance
 	 * @param sgserializable_name
 	 * @param generic

--- a/src/shogun/distance/LevenshteinDistance.h
+++ b/src/shogun/distance/LevenshteinDistance.h
@@ -114,7 +114,15 @@ namespace shogun
 
 		std::shared_ptr<Features>
 		replace_lhs(std::shared_ptr<Features> lhs) override;
+		
+		using Distance::distance;
 
+		int32_t distance(const std::string& lhs, const std::string& rhs)
+		{
+			const SGVector<char> l(lhs.begin(), lhs.end());
+			const SGVector<char> r(rhs.begin(), rhs.end());
+			return compute_impl(l, r);
+		}
 	protected:
 		/// compute distance function for features a and b
 		/// idx_{a,b} denote the index of the feature vectors

